linera-core/src/client/mod.rs

```diff
@@ -1138,78 +1144,105 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut certificates = Vec::new();
+        let mut certificates: Vec<Certificate> = Vec::new();
         let mut new_tracker = tracker;
+
         for entry in info.requested_received_log {
             let query = ChainInfoQuery::new(entry.chain_id)
                 .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+
             let local_response = self
                 .client
                 .local_node
                 .handle_chain_info_query(query.clone())
                 .await
                 .map_err(|error| NodeError::LocalNodeQuery {
                     error: error.to_string(),
-                })?;
-            if !local_response
-                .info
-                .requested_sent_certificate_hashes
-                .is_empty()
-            {
+                })?
+                .info;
+            if !local_response.requested_sent_certificate_hashes.is_empty() {
+                // We've already processed incoming messages for this certificate.
                 new_tracker += 1;
                 continue;
             }
 
-            let mut info = remote_node.handle_chain_info_query(query).await?;
-            let Some(certificate_hash) = info.requested_sent_certificate_hashes.pop() else {
-                break;
-            };
+            let remote_response = remote_node.handle_chain_info_query(query).await?;
+            let certificate_hash =
+                match remote_response.requested_sent_certificate_hashes.as_slice() {
+                    &[hash] => hash,
+                    [] => {
+                        warn!("Validator didn't have certificate he claimed to have.");
+                        break;
+                    }
+                    _ => {
+                        error!("Validator sent more than one certificate hash for a single block.");
```

Comment:
(We could even return an `InvalidChainInfoResponse` error here: that validator is faulty.)

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2643/commits/2588a2bb4232f2640c5fac85bcbb4ddf4cac9e02 .